src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/Infrastructure/RouteStringSyntaxDetector.cs

```diff
@@ -151,6 +165,16 @@ private static bool IsAnyStringLiteral(int rawKind)
 
         for (var node = token.Parent; node != null; node = node.Parent)
         {
+            // Searching for the language comment on all parents is expensive (GetLeadingTrivia) if the token is deeply nested.
```

Comment:
>GetLeadingTrivia [](http://example.com/codeflow?start=79&length=16)

Is the call to `GetLeadingTrivia()` expensive or is the issue that `HasLanguageComment()` and therefore `GetLeadingTrivia()` were called too many times for a concatenation of many strings?

Reply:
I don't know. I don't know how fast `GetLeadingTrivia()` should be.

It's true that `HasLanguageComment` was doing unneccessary work and calling `GetLeadingTrivia()` a lot.

One thing I noticed: even after the optimization in this PR to reduce calls to `HasLanguageComment`, which reduced analysis time from 5 seconds to 0.2 seconds, `GetLeadingTrivia()` is still the hottest method in the profile.

Reply:
After the optimization:
![image](https://github.com/dotnet/aspnetcore/assets/303201/d193daca-e05b-4ee7-9984-371e91ddc5f8)

Reply:
I just improved the time some more by looking ahead when matching a previously matched node. Reduces the number of calls to GetLeadingTrivia.